Add RUBY_GC_HEAP_INIT_SIZE_%d_SLOTS to pre-init pools granularly

casperisfine opened the pull request:
The old RUBY_GC_HEAP_INIT_SLOTS isn't really usable anymore as it initialize all the pools by the same factor, but it's unlikely that pools will need similar sizes.

In production our 40B pool is 5 to 6 times bigger than our 80B pool.

cc @peterzhu2118 @eightbitraptor @jasonhl

peterzhu2118 commented on gc.c:
```diff
@@ -11822,7 +11826,10 @@ ruby_gc_set_params(void)
+        gc_set_initial_pages(objspace, TRUE);
+    }
+    else {
+        gc_set_initial_pages(objspace, FALSE);
```
Why is this needed? The setup should have been done in `Init_heap` so we don't need to do it again.

casperisfine replied:
It's so that if you don't set `RUBY_GC_HEAP_INIT_SLOTS` but do set any `RUBY_GC_HEAP_INIT_SIZE_%d_SLOTS` we'll set initial pages.

peterzhu2118 reviewed (commented):
Can you also add some tests for this?

nobu commented on gc.c:
```diff
@@ -11748,15 +11748,19 @@ get_envparam_double(const char *name, double *default_value, double lower_bound,
 
     for (int i = 0; i < SIZE_POOL_COUNT; i++) {
         rb_size_pool_t *size_pool = &size_pools[i];
+        char env_key[40];
```
The necessary size can be calculated by `sizeof("RUBY_GC_HEAP_INIT_SIZE_" "_SLOTS") + DECIMAL_SIZE_OF_BITS(sizeof(size_pool->slot_size) * CHAR_BIT)`.

casperisfine commented:
> Can you also add some tests for this?

Yeah definitely, I should have opened this as a draft. I'd like to test this in prod before even considering to merge it.

peterzhu2118 commented on gc.c:
```diff
@@ -11748,15 +11748,19 @@ get_envparam_double(const char *name, double *default_value, double lower_bound,
         rb_size_pool_t *size_pool = &size_pools[i];
+        char env_key[sizeof("RUBY_GC_HEAP_INIT_SIZE_" "_SLOTS") + DECIMAL_SIZE_OF_BITS(sizeof(size_pool->slot_size) * CHAR_BIT)];
+        snprintf(env_key, sizeof(env_key), "RUBY_GC_HEAP_INIT_SIZE_%d_SLOTS", size_pool->slot_size);
+        size_t pool_init_slots = global_heap_init_slots ? gc_params.heap_init_slots : 0;
+        get_envparam_size(env_key, &pool_init_slots, 0);
```
`pool_init_slots` is being overwritten here so the original value is gone.

nobu replied:
It is only the case that env is set.

casperisfine commented:
I've cleaned up the logic and added some very basic tests. 

I could add some assertions on the pool sizes, but I fear it may be a bit fragile on other arch? @peterzhu2118 if you think it shouldn't be flaky I can add them.

peterzhu2118 commented on test/ruby/test_gc.rb:
```diff
@@ -315,6 +315,15 @@ def test_gc_parameter
     }
     assert_normal_exit("exit", "[ruby-core:39777]", :child_env => env)
 
+    env = {}
+    GC.stat_heap.each do |_, s|
+      env["RUBY_GC_HEAP_INIT_SIZE_#{s[:slot_size]}_SLOTS"] = "200000"
+    end
+    assert_normal_exit("exit", "", :child_env => env)
+
+    env["RUBY_GC_HEAP_INIT_SLOTS"] = "100000"
+    assert_normal_exit("exit", "", :child_env => env)
```
Can you also add a test to ensure that it doesn't crash with a small number (like 0)?

casperisfine replied:
Done.

peterzhu2118 commented on gc.c:
```diff
@@ -11822,7 +11830,10 @@ ruby_gc_set_params(void)
+        gc_set_initial_pages(objspace, TRUE);
+    }
+    else {
+        gc_set_initial_pages(objspace, FALSE);
```
I think this resets all of the `heap_allocatable_pages`.

This branch:

```
$ ruby --disable-gems -e "GC.stat_heap.each { |k, v| puts k.to_s + '=' + v[:heap_allocatable_pages].to_s }"
0=0
1=0
2=0
3=0
4=0
```

Master branch:

```
$ ruby --disable-gems -e "GC.stat_heap.each { |k, v| puts k.to_s + '=' + v[:heap_allocatable_pages].to_s }"
1=11
2=22
3=48
4=97
```

casperisfine replied:
Good catch, I fixed it.

rafaelfranca commented:
How would users know which are the possible values of `%d` in order to configure?

casperisfine commented:
> How would users know which are the possible values of `%d` in order to configure?

It's exposed via `GC.stat_heap`

```ruby
>> GC.stat_heap
=> 
{0=>{:slot_size=>40, ...},
 1=>{:slot_size=>80, ...},
 2=>{:slot_size=>160, ...},
 3=>{:slot_size=>320, ...},
 4=>{:slot_size=>640, ...}}
```

rafaelfranca commented:
My question is more about where is this documented. With `RUBY_GC_HEAP_INIT_SIZE_%d_SLOTS` being exposed, I believe it should be documented, so people know what is the expected values of `%d` and even where to find them.

casperisfine commented:
> My question is more about where is this documented.

Ok, `RUBY_GC_HEAP_INIT_SLOTS` is only documented in the manpage, so I added these new variables there.

casperisfine commented on man/ruby.1:
```diff
@@ -550,7 +550,22 @@ There are currently 4 possible areas where the GC may be tuned by
 .It Ev RUBY_GC_HEAP_INIT_SLOTS
-Initial allocation slots.  Introduced in Ruby 2.1, default: 10000.
+Initial allocation slots. Applies to all slot sizes.  Introduced in Ruby 2.1, default: 10000.
+.Pp
+.It Ev RUBY_GC_HEAP_INIT_SIZE_40_SLOTS
+Initial allocation of size 40 slots.  Introduced in Ruby 3.3.
+.Pp
+.It Ev RUBY_GC_HEAP_INIT_SIZE_80_SLOTS
+Initial allocation of size 80 slots.  Introduced in Ruby 3.3.
+.Pp
+.It Ev RUBY_GC_HEAP_INIT_SIZE_160_SLOTS
+Initial allocation of size 160 slots.  Introduced in Ruby 3.3.
+.Pp
+.It Ev RUBY_GC_HEAP_INIT_SIZE_320_SLOTS
+Initial allocation of size 320 slots.  Introduced in Ruby 3.3.
+.Pp
+.It Ev RUBY_GC_HEAP_INIT_SIZE_640_SLOTS
+Initial allocation of size 640 slots.  Introduced in Ruby 3.3.
```
Hum, right. Should I only document `RUBY_GC_HEAP_INIT_SIZE_%d_SLOTS` and explain how to list the pool sizes?

peterzhu2118 replied:
Yeah I think that would be a good way to document it.

mttkay commented:
@casperisfine I was wondering, are there any stats available for size pool occupation (i.e. live objects per pool?)

I am not sure how one would now calculate total memory use. Previously this was just the 40B multiple of `GC.stat[:heap_live_slots]` plus whatever memory was malloc'ed for any objects that did not fit a single slot.

However, now objects are distributed over a number of pools, and `GC.stat_heap` only returns sizing stats, but does not include information on how many live objects are in each pool (which we could then multiple with the slot size of the respective pool to arrive at total memory use per pool?)

Thank you!

byroot commented:
> I was wondering, are there any stats available for size pool occupation (i.e. live objects per pool?)

Yes, it's in `GC.stat_heap`.

mttkay commented:
You mean `heap_eden_slots`?

byroot commented:
Yes.

mttkay commented:
@byroot sorry for my ignorance, but would this actually give me actual memory consumed by the application? Was just looking over https://scoutapm.com/blog/copy-manage-ruby-memory-usage again, and the way I understand eden slots to work is that they _may_ hold a live object, but may also be tagged free. Of course they still consume process memory, but that cannot be attributed to the application if it's unused?

So this is still different from the original `heap_live_slots`, which are only those slots actually occupied by an instance of RVALUE that represents an actual live object? :thinking: I might be misunderstanding how the GC manages these.

mttkay commented:
Asked differently: I am looking for the per-pool equivalent of:

```
heap_live_slots

    The total number of slots which contain live objects

heap_free_slots

    The total number of slots which do not contain live objects
```

https://docs.ruby-lang.org/en/master/GC.html#method-c-stat

Are you saying `heap_eden_slots =~ heap_live_slots` (just for a given pool)? What is the equivalent of `heap_free_slots` for a pool in `stat_heap`?

mttkay commented:
This doesn't sound right @byroot. When booting into an irb:

```pry
irb(main):002> GC.stat[:heap_live_slots]
=> 123976
irb(main):005> GC.stat_heap.each_value.sum { |pool| pool[:heap_eden_slots] }
=> 332241
```

Total live slots across all heaps are about 1/3 of all reported eden slots in pools. So I would conclude these metrics measure different things?

byroot commented:
Yeah, sorry. `heap_eden_slot` include all slots that are allocated (from an OS point of view) regardless of whether the slot is in use.

> but would this actually give me actual memory consumed by the application? 

If that's the answer you are looking for, then slots aren't really what you should look for. You're better to use `ObjectSpace.dump_all` and sum all the `memsize`, that's part of what https://github.com/Shopify/heap-profiler does.

Now of course this is rather slow, so I don't know if you are trying to do this at runtime with a minimal overhead, in which case, I don't have a solution for you. Even if you had the proper number of slots in use, with their respective sizes, that would still overlook all the memory that is allocated via `malloc`, which is likely an extremely large portion.

mttkay commented:
I see, thank you.

Yes, we are doing this at runtime using an MRI patch that tracks per-thread heap allocations. Previously, it was then trivial (and cheap) to compute per-request memory use like so:

```
total_memory_used = (total_live_objects * RVALUE_SIZE) + large_obj_malloced_bytes
```

The last component, malloc'ed bytes for objects that don't fit an RVALUE slot were tracked in the VM directly.

However, this calculation is now wrong (it underestimates live memory), because we don't know anymore how `total_live_objects` distributes across the GC-managed heap (object space), which now all have different slot sizes. :thinking:
